### PR TITLE
Refactoring domain

### PR DESCRIPTION
**Pull request by basz**
I don't see other business logic candidates to move to the domain model. you?

### CODE REVIEW

File: src/Server/AuthorizationServer.php [hunk 1]
```diff
@@ -312,7 +312,7 @@ private function getClient(ServerRequestInterface $request, $allowPublicClients)
         $client = $this->clientService->getClient($id);
 
         // We delegate all the checks to the client service
-        if (null === $client || (!$allowPublicClients && !$this->clientService->authenticate($client, $secret))) {
+        if (null === $client || (!$allowPublicClients && !$client->authenticate($client, $secret))) {
```
Comment (bakura10) [inline src/Server/AuthorizationServer.php]: This feels strange. This should be $client->authenticate($secret).

File: tests/Server/Service/ClientServiceTest.php [hunk 2]
```diff
@@ -71,8 +71,8 @@ public function testRegisterClient()
         $this->assertEquals(60, strlen($client->getSecret()));
         $this->assertEquals(40, strlen($secret));
 
-        $this->assertFalse($this->clientService->authenticate($client, 'azerty'));
-        $this->assertTrue($this->clientService->authenticate($client, $secret));
-        $this->assertFalse($this->clientService->authenticate($client, $client->getSecret()));
+        $this->assertFalse($client->authenticate('azerty'));
```
Comment (bakura10) [inline tests/Server/Service/ClientServiceTest.php]: Those tests should be moved to the ClientTest (the model).

**Comment by bakura10** [comment]
As said, I think you should remove all setter, and enforce setting parameters directly into the constructor :).

**Comment by bakura10** [comment]
Just need to do the same for the Token mode :).

**Comment by basz** [comment]
easier said then done… :-/

It is not clear to me what the signature must be of the __construct for AbstractToken.

The first one ‘token' is clear, not optional...

client and owner seem optional too looking at https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Grant/AbstractGrant.php#L52

scopes are not optional but could be

expiresAt is optional for Refresh- and AccessTokens : https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Entity/RefreshToken.php#L34

also AuthorizationCode has an extra parameter which is not optional and should therefore not be simply added as last argument to the constructor.

https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Entity/AuthorizationCode.php#L35

This would mean the constructor signature for tokens are different per kind, which makes it harder to create an instance in AbstractToken. 
Currently a token is instanciated in the different Grants and modified as needed (via setters) in token service before storage. https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Service/TokenService.php#L92

I’m not sure a token is/should be an immutable value object. Some sort of factory method on the services as we need access to the ttl value and repository, perhaps...

Will this blow up in our face if we decide to add support for other kinds of tokens such as JWT or Mac?

pfft. i need to think on this. jump in if you feel you can help

**Comment by bakura10** [comment]
Yeah, I agree that this is a bit hard. Actually the idea (I've learned that from the awesome @danizord) is to make model useful, and not just stupid objects.

For instance, all those:

https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Service/TokenService.php#L108-L112

Could be moved to the AbstractToken constructor. Which already removes one setter (setToken). This actually makes sense, because the token is actually the primary key, but the setToken would allow to change it, which should be prohibited.

You are then only left with three setters (setExpiresAt, setClient and setOwner), that I think can't be removed.

I'm sure there are better ways to do it and make the model even more useful (maybe @danizord has some ideas), but for now I think it'll be enough.

**Comment by basz** [comment]
> Could be moved to the AbstractToken constructor

not really as the Token Model should/will not have access the tokenRepository.

> You are then only left with three setters (setExpiresAt, setClient and setOwner), that I think can't be removed.

and setScope
and setRedirectUri for AuthorisationCodeTokens

What I now have done is this;

each Grant collects data relevant for token creation into an array and passes that to createToken along with the token class it is expecting. (would be easier if we had concrete instances of the token services)

```
AbstractToken::createToken(AbstractToken $token): AbstractToken
```
is changed to
```
AbstractToken::createToken(string $tokenClass, array $tokenData): AbstractToken
```
AbstractToken::createToken generates and validates the last pieces of the data and only then WHEN ALL DATA IS READY we construct either an AuthorisationCode token or Access/Refresh token based on $tokenClass. No need to mutate at all.

(https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Service/TokenService.php#L33)[consider this] I think concrete instances of the TokenServices would be useful. An accessTokenService should only concern itself with AccessToken. Especially since I have no idea at this time how other Token (jwt) types might need to be constructed.

**Comment by bakura10** [comment]
Here: https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Grant/PasswordGrant.php#L122-L125

What about modifyig this logic and having this in your grant:

```php
$accessToken = $this->tokenService->generateNew($scopes, $owner, $client);
```

The service would then do something like:

```php
public function generateNew(array $scopes, TokenOwnerInterface $owner = null, Client $client = null): AccessToken
{
    if (empty($scopes)) {
        $scopes = $this->getDefaultScopes();
    }

    $token = new AccessToken($this->getTtl(), $scopes, $owner, $client)   
}
```

Which gives all the info for our constructor:

```php
public function __construct(int $ttl, array $scopes, TokenOwnerInterface $owner = null, Client $client = null)
{
    $this->token     = bin2hex(random_bytes(20));
    $this->expiresAt = (new DateTimeImmutable())->modify('+ ' + $ttl + ' seconds');
    $this->owner     = $owner;
    $this->client    = $client;
}
```

As said, it may make sense to have specialize service classes for access, authorization and refresh token, so that the generate new may accept different parameters.

Does it make a bit of sense? :)

**Comment by basz** [comment]
what about this check?

https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Service/TokenService.php#L110

what happens to the token value here, constructor will run again...

https://github.com/zf-fr/zfr-oauth2-server/blob/master/src/Server/Service/TokenService.php#L126

**Comment by bakura10** [comment]
What about something like that?

```php
public function generateNew(array $scopes, TokenOwnerInterface $owner = null, Client $client = null): AccessToken
{
    if (empty($scopes)) {
        $scopes = $this->getDefaultScopes();
    }
   
    do {
       $token = new AccessToken($this->getTtl(), $scopes, $owner, $client);  
    } while ($this->tokenRepository->tokenDoesNotExist($token->getToken());

    $this->tokenRepository->save($token);
}
```

**Comment by basz** [comment]
about to give this a second attempt...

tokenDoesExist will work, no hydration is required. But there are three invocations of were a returned Token is expected.

```
$this->xxxxTokenService->getToken($token);

public function getToken(string $token) {
    $tokenFromDb = $this->tokenRepository->findByToken($token);

   // $tokenFromDb needs to be a hydrated object
   // so both the token and the expiresAt values must come from the persistence layer... see the problem?
}
``` 
Your proposed constructor would work for NEW tokens, but not for new instances that need to encapsulated persist data;
```
public function __construct(int $ttl, array $scopes, string $token = null, TokenOwnerInterface $owner = null, Client $client = null)
{
    // this would work for both new and persisted tokens
    $this->token     = $token ?? bin2hex(random_bytes(20));
    // for a new token $ttl would work, but for the expiresAt column not.
    $this->expiresAt = (new DateTimeImmutable())->modify("+$ttl seconds");
```

**Comment by bakura10** [comment]
For "hydration" what we do on our project is creating a static method called "reconstitute" (we found the naming pretty good).

Actually the Doctrine repository won't really need it, because Doctrine already does hydration under the hood using reflection, but typically the reconstitute method works like that:

```php
public function reconstitute(array $data): Token
{
    $token = new self();
    $token->id = $data['id'];
    // ...
}
```

Therefore each repository that does not have a auto-hydration mechanism, can work like this:

```php
public function findTokenById(string $id): Token
{
     $data = // Do you call to Database

     return Token::reconstitute($data);
}


As a consequence, what we did so far in our project is making the constructor private and not accepting any parameter. It's then up to the specialized static method to construct the object. In our case we would simply have the `generateNew` that creates a new token, and the `reconstitute` that takes raw array data, and populate the token from the incoming data. This `reconstitute` method will be used by any persistence repository that does not have any auto-hydration.

Does that make sense? :)

File: src/Server/Model/AbstractToken.php [hunk 3]
```diff
@@ -57,87 +58,119 @@
...
-     * @param  Client $client
-     * @return void
+     * @param array $data
+     * @return AbstractToken
      */
-    public function setClient(Client $client)
+    public static function reconstitute(array $data)
     {
-        $this->client = $client;
+        $data['owner']     = $data['owner'] ?? null;
```
Comment (bakura10) [inline src/Server/Model/AbstractToken.php]: I'm not sure why you're doing this. You should direclty assign to the object properties, this extra step seems useless.

Comment (basz) [inline src/Server/Model/AbstractToken.php]: $data['scopes'] might not be set. but i should move it to the object props as you say.

**Comment by basz** [comment]
last update only done on the domain. 

- [x] private properties, 
- [x] no setters
- [x] specialised methods for createNewSomething. Note that I had to use AccessToken::createNewAccessToken, RefreshToken::createNewRefreshToken, etc. as apparently (i was told) "you are violating L in SOLID - objects in a program should be replaceable with instances of their subtypes without altering the correctness of that program" and "if it takes an extra argument, it does something extra, so it should be named accordingly".
- [x] methods reconstitute, for (optional) hydration if the arm does not do that for ya

- [x] fully tested, 100% coverage

**Comment by bakura10** [comment]
Tests are not passing :)

**Comment by basz** [comment]
Model tests are passing. I need to hear from THE Grant master before i incorporate the model into the rest. Pls only look at the model structure for now.

**Comment by basz** [comment]
i believe this branch is about done. when reviewed and merged I'll continue with #44. after which i'll start on the doctrine module.

File: src/Server/Grant/RefreshTokenGrant.php [hunk 4]
```diff
@@ -53,10 +54,10 @@ class RefreshTokenGrant extends AbstractGrant
     private $rotateRefreshTokens = false;
 
     /**
-     * @param TokenService $accessTokenService
-     * @param TokenService $refreshTokenService
+     * @param AccessTokenService  $accessTokenService
+     * @param RefreshTokenService $refreshTokenService
      */
-    public function __construct(TokenService $accessTokenService, TokenService $refreshTokenService)
+    public function __construct(TokenService $accessTokenService, RefreshTokenService $refreshTokenService)
```
Comment (bakura10) [inline src/Server/Grant/RefreshTokenGrant.php]: Typehint on AccessTokenService?

File: src/Server/Model/AbstractToken.php [hunk 5]
```diff
@@ -39,105 +40,126 @@
...
+        }
+
+        if (is_array($scopes)) {
+            foreach ($scopes as &$scope) {
+                $scope = $scope instanceof Scope ? $scope->getName() : (string) $scope;
+            }
+        }
+
+        $class = get_called_class();
+        $token = new $class();
```
Comment (bakura10) [inline src/Server/Model/AbstractToken.php]: I didn't try but did you try to do $token = new static(); ? I didn't try it should be worth the try :).

File: src/Server/Repository/RefreshTokenRepositoryInterface.php [hunk 6]
```diff
@@ -18,9 +18,16 @@
 
 namespace ZfrOAuth2\Server\Repository;
 
+use ZfrOAuth2\Server\Model\RefreshToken;
+
 /**
  * Interface RefreshTokenRepositoryInterface
  */
 interface RefreshTokenRepositoryInterface extends TokenRepositoryInterface
```
Comment (bakura10) [inline src/Server/Repository/RefreshTokenRepositoryInterface.php]: I'm not sure it makes sense now to have a TokenRepositoryInterface considering that we have specialized interface. Then you would be able to typehint other methods (deleteToken, tokenExists...) using the concrete model.

Comment (bakura10) [inline src/Server/Repository/RefreshTokenRepositoryInterface.php]: Never mind, I understand the reasn: it allows you tos implify the TokenService class by having the same constructor.

File: src/Server/Service/TokenService.php [hunk 7]
```diff
@@ -18,47 +18,34 @@
...
- * as the token TTL)
- *
- * @TODO    : should we create one service per token type? I think it's a bit useless, as the only thing that would
- *        be overridden is the token TTL
- *
  * @author  Michaël Gallego <mic.gallego@gmail.com>
  * @licence MIT
  */
-class TokenService
+abstract class TokenService
```
Comment (basz) [inline src/Server/Service/TokenService.php]: Should we rename this class to `AbstractTokenService`?

Comment (bakura10) [inline src/Server/Service/TokenService.php]: Definitely

**Comment by basz** [comment]
superseded by #44

**Pull request closed without merging**